DAOS-4834 test: Enable secure mode for daos_server, daos_agent, and dmg tools for security tests on CI.

dinghwah opened the pull request:
Current 16 failed tests with certificates enabled will be running with insecure-mode, that will be addressed on the following JIRA tasks.

Signed-off-by: Ding Ho <ding-hwa.ho@intel.com>

daosbuild1 commented on src/tests/ftest/util/dmg_utils.py:
```diff
@@ -1071,7 +910,7 @@ def storage_format(path, hosts, insecure=True):
     dmg = DmgCommand(path)
     dmg.insecure.value = insecure
-    dmg.hostlist.value = hosts
+    dmg.hostlist.value = host
```
(pylint-undefined-variable) Undefined variable 'host'

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/1/execution/node/58/log

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/2/execution/node/546/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/2/execution/node/642/log

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/2/execution/node/807/log

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/3/execution/node/580/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/3/execution/node/610/log

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-2980/3/testReport/(root)/

dinghwah commented:
(1)PR to update launch.py with optional -sec (secure-mode) to turn on 11 functional tests with certificates. 
(2)There is no impact to the rest of the functional tests.
(3)The failures on CentOS 7 RPMs is track by DAOS-5207.

brianjmurrell commented:
> (3)The failures on CentOS 7 RPMs is track by DAOS-5207.

To be clear, this cannot land until DAOS-5207 is resolved, otherwise we are adding another "known failure" to the pile.

brianjmurrell commented on src/tests/ftest/launch.py:
```diff
@@ -708,10 +732,25 @@ def replace_yaml_file(yaml_file, args, tmp_dir):
+    # Generate certificates if not exist
+    print(" ")
+    if not os.path.exists("./daosCA"):
+        subprocess.call(["../../../../../utils/certs/gen_certificates.sh"])
```
Will this path be correct if installed by RPMs?  I suspect not.  Your probably want to base this path on `PREFIX`.

You can test in CI with RPMs by adding `RPM-test: true` to your commit message.

dinghwah replied:
Secure by default is the final goal. 
We want to turn on the certificates for our functional tests gradually, or the failures will be unpredictable and overwhelm.
Currently, all the certificates tests were done manually, I choose 11 functional tests as pilot-test with certificates for CI.
At this phase, the location of the certificates is not critical.
We want to see the stability of the 11 functional tests with certificates.

brianjmurrell replied:
Even if we want to enable it gradually (I assume you mean over time, even once the initial patch lands), the patch that lands (i.e. this patch) that enables the feature should still be **secure by default** and the tests that you don't want to enable security on should have it specifically disabled, rather than the opposite stance that this patch is taking.

dinghwah replied:
When we turn on the secure-mode on functional tests, we want to make sure the tests are validated and stabilized, that's the pilot tests with certificates for.

brianjmurrell replied:
So to be clear, you are not seeking to land this PR as is, using **insecure by default** mode, is that correct?

dinghwah replied:
Insecure-mode is currently in default, and this PR is to enable the certificates with 11 pilot tests for CI.

brianjmurrell replied:
OK.  I will maintain my -1 at least until **secure by default** is enabled.

brianjmurrell commented on src/tests/ftest/launch.py:
```diff
@@ -708,10 +732,25 @@ def replace_yaml_file(yaml_file, args, tmp_dir):
+
+    """
+    # Generate certificates if not exist
+    print(" ")
```
Why is this needed?

brianjmurrell commented on src/tests/ftest/launch.py:
```diff
@@ -708,10 +732,25 @@ def replace_yaml_file(yaml_file, args, tmp_dir):
+    print(" ")
+    if not os.path.exists("./daosCA"):
+        subprocess.call(["../../../../../utils/certs/gen_certificates.sh"])
+    print(" ")
```
Why is this needed?

brianjmurrell commented on src/tests/ftest/launch.py:
```diff
@@ -1281,6 +1320,10 @@ def main():
         help="limit output to pass/fail")
+    parser.add_argument(
+        "-sec", "--secure_mode",
+        action="store_true",
```
My recollection of _argparse_ is fuzzy, but doesn't `store_true` mean the default value for the option is `False`?  We should be **secure by default**.  That's an SDL principle.

dinghwah replied:
Please see my comment above.
In this phase, -sec as an option.

brianjmurrell replied:
It shouldn't be.  _Insecure_ should be the option.  _Secure_ should be the default.

dinghwah replied:
The secure-mode by default will be after the pilot tests stabilized in the CI.

brianjmurrell commented on src/tests/ftest/security/pool_connect_init.yaml:
```diff
@@ -7,7 +7,7 @@ server_config:
   mode: 511
   name: daos_server
-  scm_size: 16777216
+  scm_size: 136777216
```
Was a `3` accidentally inserted in this value?  If not, why is the `scm_size` being increased here?

```suggestion
  scm_size: 16777216
```

brianjmurrell replied:
Same comments as the other `scm_size` increase.

dinghwah replied:
scm_size of 16777216 is too small and causing error pool creation.

brianjmurrell replied:
Perhaps, but please see [my comments](https://github.com/daos-stack/daos/pull/2980#discussion_r446164608) on the other `scm_size` change.

dinghwah replied:
All scm_size changes in test.yaml are due to the same reason.

brianjmurrell commented on src/tests/ftest/security/pool_security_groups.yaml:
```diff
@@ -5,24 +5,15 @@ hosts:
-     insecure: False
 pool_acl:
-  scm_size: 134217728
+  scm_size: 1334217728
```
Was a `3` accidentally inserted in this value? If not, why is the `scm_size` being increased here?

```suggestion
  scm_size: 134217728
```

dinghwah replied:
Change the scm_size due to origin scm_size 134217728 too small and causing failure.

brianjmurrell replied:
Is turning on security actually making the SCM usage higher?  In the most recently fully run weekly test, this test passed all of it's tests:

https://build.hpdd.intel.com/job/daos-stack/job/daos/job/weekly-testing/86/testReport/(root)/DaosRunPoolSecurityTest/

So are you sure any failure you are seeing is related to SCM size?

How as the new value chosen?  It appears to be just an insertion of a `3` into the existing value increasing it by 9.94x.  Surely something more scientific should be done to choose a reasonable new value, yes?

Also, please don't resolve conversations that you didn't start.  The Resolve button is for the person who raised the issue to acknowledge that the issue has been resolved to their satisfaction.

dinghwah replied:
From the failure, the min scm_size is about 126M.

brianjmurrell replied:
Any idea why this was working as recently as https://build.hpdd.intel.com/job/daos-stack/job/daos/job/weekly-testing/86/testReport/(root)/DaosRunPoolSecurityTest/ then?

brianjmurrell commented on src/tests/ftest/security/pool_security_groups.yaml:
```diff
@@ -5,24 +5,15 @@ hosts:
   user_prefix: daos_ci
-  num_user: 1
-  num_group: 1
+  num_user: 5
+  num_group: 5
```
Is this change needed in order to enable secure mode?

dinghwah replied:
no

brianjmurrell replied:
So it needs to be fixed before the conversation can be resolved -- by the person who started the conversation, once they have confirmed that it as been fixed.

brianjmurrell commented on src/tests/ftest/util/agent_utils_params.py:
```diff
@@ -33,14 +33,12 @@ class DaosAgentTransportCredentials(TransportCredentials):
+            "/run/transport_config/*", "transport_config")
 
+        self.allow_insecure = BasicParameter(False, False)
+        self.ca_cert = BasicParameter(None, "./daosCA/certs/daosCA.crt")
+        self.cert = BasicParameter(None, "./daosCA/certs/agent.crt")
+        self.key = BasicParameter(None, "./daosCA/certs/agent.key")
```
These paths likely will not work with RPM testing.  As mentioned before you should test this PR with "RPM-test: true" in your final commit also to ensure you are not breaking testing-with-RPMs.

brianjmurrell replied:
I don't understand why you resolved this conversation without doing anything to address it.  This is why it's the responsibility of the person who started the conversation to resolve them.

brianjmurrell commented on src/tests/ftest/util/apricot/apricot/test.py:
```diff
@@ -274,6 +274,7 @@ def __init__(self, *args, **kwargs):
             os.path.split(self.filename)[1], self.name.str_uid)
         # self.debug = False
         # self.config = None
+        self.insecure_mode = True
```
This looks like it's not **secure by default**.

dinghwah replied:
It is correct, insecure by default with optional -sec for the secure mode.

brianjmurrell replied:
See previous comments about **secure by default**.

brianjmurrell commented on src/tests/ftest/util/command_utils_base.py:
```diff
@@ -477,20 +477,6 @@ def __init__(self, namespace, title):
-        #   - key: <file>, e.g. ".daos/daos_agent.key"
-        #       Key portion of Server Certificate
-        #
         self.allow_insecure = BasicParameter(True, True)
```
Should this be changed to **secure by default**?

dinghwah replied:
same as above.

brianjmurrell replied:
Ditto.

brianjmurrell commented on src/tests/ftest/util/dmg_utils_params.py:
```diff
@@ -32,7 +32,12 @@ class DmgTransportCredentials(TransportCredentials):
+            "/run/transport_config/*", "transport_config")
+
+        self.allow_insecure = BasicParameter(False, False)
+        self.ca_cert = BasicParameter(None, "./daosCA/certs/daosCA.crt")
+        self.cert = BasicParameter(None, "./daosCA/certs/admin.crt")
+        self.key = BasicParameter(None, "./daosCA/certs/admin.key")
```
Ditto with the paths and testing with RPMs.

brianjmurrell commented on src/tests/ftest/util/server_utils_params.py:
```diff
@@ -35,14 +35,13 @@ class DaosServerTransportCredentials(TransportCredentials):
+            "/run/transport_config/*", "transport_config")
 
+        self.allow_insecure = BasicParameter(False, False)
+        self.client_cert_dir = BasicParameter(None, "./daosCA/certs")
+        self.ca_cert = BasicParameter(None, "./daosCA/certs/daosCA.crt")
+        self.cert = BasicParameter(None, "./daosCA/certs/server.crt")
+        self.key = BasicParameter(None, "./daosCA/certs/server.key")
```
Paths again.

dinghwah commented on src/tests/ftest/launch.py:
```diff
@@ -614,6 +614,20 @@ def replace_yaml_file(yaml_file, args, tmp_dir):
 
+        #Check for transport_config and allow_insecure from the test yaml
+        trans_conf_defined_in_yaml = False
+        if "transport_config" in yaml_data.keys():
```
Has to check transport_config as key, since transport_config could be on 2nd or 3rd layer of the dictionary.

dinghwah commented on src/tests/ftest/launch.py:
```diff
@@ -686,14 +700,24 @@ def replace_yaml_file(yaml_file, args, tmp_dir):
             return None
 
+        if trans_conf_defined_in_yaml:
+            if not bool(allow_insec_yaml):
```
I saw test failed w/o bool.

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/4/execution/node/620/log

phender commented on src/tests/ftest/util/dmg_utils_params.py:
```diff
@@ -32,7 +32,12 @@ class DmgTransportCredentials(TransportCredentials):
+        self.allow_insecure = BasicParameter(False, False)
+        self.ca_cert = BasicParameter(None, "./daosCA/certs/daosCA.crt")
+        self.cert = BasicParameter(None, "./daosCA/certs/admin.crt")
+        self.key = BasicParameter(None, "./daosCA/certs/admin.key")
```
Sorry I missed this something earlier.  If now I'm seeing this correctly there are different values for `cert` and `key` depending on if we are setting up the `DaosServerTransportCredentials`, `DaosAgentTransportCredentials`, or the `DmgTransportCredentials`.  In other words:

- For `daos_server` the 'server' files are used:
```
        self.cert = BasicParameter(None, "./daosCA/certs/server.crt")
        self.key = BasicParameter(None, "./daosCA/certs/server.key")
```
- For `daos_agent` the 'admin' files are used:
```
        self.cert = BasicParameter(None, "./daosCA/certs/agent.crt")
        self.key = BasicParameter(None, "./daosCA/certs/agent.key")
```
- For `dmg` the 'admin' files are used:
```
        self.cert = BasicParameter(None, "./daosCA/certs/admin.crt")
        self.key = BasicParameter(None, "./daosCA/certs/admin.key")
```

The `allow_insecure` and `ca_cert` do appear to be common thought, correct?

Assuming this is true, I'll amend my previous comments and say that the all three `DaosServerTransportCredentials`, `DaosAgentTransportCredentials`, or the `DmgTransportCredentials classes should remain, but their `__inti__()` methods should only define unique default values for the `cert` and `key` values - even then they should probably use `BasicParameter.update_default(<value>)` after the `super().__init__()`.  The defaults for the common `allow_insecure` and `ca_cert` are better off defined in `TransportCredentials.__init__()`.

dinghwah replied:
Keep all the credential information on server, agent and dmg in on place would be easier for later maintenance and debug, rather than break into 2 places. 
ie:   separate 
        self.allow_insecure = BasicParameter(False, False)
        self.ca_cert = BasicParameter(None, "./daosCA/certs/daosCA.crt")     
and
        self.cert = BasicParameter(None, "./daosCA/certs/server.crt")
        self.key = BasicParameter(None, "./daosCA/certs/server.key")

dinghwah replied:
(2)gen_certificates.sh only run once not on all the hosts.
(3)When transport_config defined in the test.yaml, launch.py will NOT replace the allow_insecure value, even a different -sec mode is specified in the launch.py.

dinghwah commented on src/tests/ftest/launch.py:
```diff
@@ -708,10 +732,25 @@ def replace_yaml_file(yaml_file, args, tmp_dir):
+    """
+    # Generate certificates if not exist
+    print(" ")
+    if not os.path.exists("./daosCA"):
+        subprocess.call(["../../../../../utils/certs/gen_certificates.sh"])
```
No. 
This shell program only run once, does not run on all host.

daosbuild1 commented:
Test stage Functional_Hardware_Small completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-2980/4/testReport/(root)/

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/4/execution/node/760/log

daosbuild1 commented on src/tests/ftest/util/apricot/apricot/test.py:
```diff
@@ -773,7 +784,7 @@ def get_dmg_command(self, index=0):
         dmg_config_file = self.get_config_file("daos", "dmg")
         dmg_cfg = DmgYamlParameters(
-            dmg_config_file, self.server_group, DmgTransportCredentials())
+            dmg_config_file, self.server_group, DmgTransportCredentials(self.workdir))
```
(pylint-line-too-long) Line too long (86/80)

daosbuild1 commented on src/tests/ftest/util/apricot/apricot/test.py:
```diff
@@ -343,6 +344,9 @@ def setUp(self):
             hosts.extend(self.hostlist_clients)
         self.stop_leftover_processes(["orterun"], hosts)
+        self.insecure_mode = self.params.get("allow_insecure",
+                                             "/run/server_config/transport_config/*")
```
(pylint-line-too-long) Line too long (85/80)

daosbuild1 replied:
(pylint-line-too-long) Line too long (85/80)

daosbuild1 commented on src/tests/ftest/util/server_utils_params.py:
```diff
@@ -32,16 +32,35 @@ class DaosServerTransportCredentials(TransportCredentials):
         """Initialize a TransportConfig object."""
         super(DaosServerTransportCredentials, self).__init__(
-            "/run/server_config/transport_config/*", "transport_config")
+            "/run/server_config/transport_config/*", "transport_config", log_dir)
```
(pylint-line-too-long) Line too long (81/80)

daosbuild1 commented on src/tests/ftest/util/dmg_utils_params.py:
```diff
@@ -21,18 +21,24 @@
   Any reproduction of computer software, computer software documentation, or
   portions thereof marked with this legend must also reproduce the markings.
 """
+import os
```
(pylint-unused-import) Unused import os

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/5/execution/node/58/log

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/6/execution/node/59/log

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/6/execution/node/563/log

daosbuild1 commented:
Test stage Functional_Hardware_Small completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-2980/6/display/redirect

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/7/execution/node/528/log

daosbuild1 commented:
Test stage Functional_Hardware_Small completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/7/execution/node/632/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/7/execution/node/685/log

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/7/execution/node/724/log

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/7/execution/node/803/log

daosbuild1 commented on src/tests/ftest/util/apricot/apricot/test.py:
```diff
@@ -475,9 +476,9 @@ def add_agent_manager(self, config_file=None, common_cfg=None, timeout=30):
-                self.server_group, DaosAgentTransportCredentials())
-
+            agent_transport = DaosAgentTransportCredentials(self.workdir)
+            agent_transport.allow_insecure.value = self.agent_insecure_mode
```
(pylint-no-member) Instance of 'TestWithServers' has no 'agent_insecure_mode' member

daosbuild1 replied:
(pylint-no-member) Instance of 'TestWithServers' has no 'agent_insecure_mode' member

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/8/execution/node/58/log

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/8/execution/node/595/log

daosbuild1 commented:
Test stage Functional_Hardware_Small completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/8/execution/node/675/log

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/8/execution/node/664/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/8/execution/node/686/log

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/8/execution/node/805/log

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/9/execution/node/58/log

phender commented on src/tests/ftest/util/command_utils_base.py:
```diff
@@ -476,25 +487,11 @@ def __init__(self, namespace, title):
+        default_insecure = default_insecure.lower() == "true"
+        self.ca_cert = LogParameter(log_dir, None, "daosCA.crt")
+        self.allow_insecure = BasicParameter(None, default_insecure)
+        print("[DEBUG] default_insecure= ", default_insecure)
```
Shouldn't we remove the debug statement?
```suggestion
```

phender commented on src/tests/ftest/util/command_utils_base.py:
```diff
@@ -513,6 +510,56 @@ def get_yaml_data(self):
+
+        Returns:
+            data (dict): a dictionary of parameter directory name keys and
+            value.
```
```suggestion
                value.
```

phender commented on src/tests/ftest/util/command_utils_base.py:
```diff
@@ -513,6 +510,56 @@ def get_yaml_data(self):
+        """Get certificate data by name_list.
+
+        Args:
+            name_list (list): name list.
```
```suggestion
            name_list (list): list of certificate attribute names
```

phender commented on src/tests/ftest/util/command_utils_base.py:
```diff
@@ -513,6 +510,56 @@ def get_yaml_data(self):
+        return data
+
+    def copy_certificates(self, source, hosts):
+        """ copy certificates file to the dest hosts
```
```suggestion
        """Copy certificates files from the source to the destination hosts.
```

phender commented on src/tests/ftest/util/command_utils_base.py:
```diff
@@ -513,6 +510,56 @@ def get_yaml_data(self):
+
+        Args:
+            source (str): source of the certificate files.
+            hosts (list of string): list of the destination hosts.
```
```suggestion
            hosts (list): list of the destination hosts.
```

phender commented on src/tests/ftest/util/command_utils_base.py:
```diff
@@ -513,6 +510,56 @@ def get_yaml_data(self):
+                run_command(
+                    "clush -S -v -w {} --copy {} --dest {}".format(
+                        ",".join(hosts), src_file, dst_file))
+            # debug to list copy of cert files
+            print(
+                "[DEBUG] listing of certificate files in {} on {}".format(
+                    name, hosts))
```
Do we want to keep debug statements?

phender commented on src/tests/ftest/util/dmg_utils.py:
```diff
@@ -130,6 +131,15 @@ def hostlist(self, hostlist):
+        """Set certificates for the hostlist.
+
+        Args:
+            hostlist (string list): list of host addresses
```
The '()' should contain a python type.
```suggestion
            hostlist (list): list of host addresses
```

phender commented on src/tests/ftest/util/dmg_utils_params.py:
```diff
@@ -22,17 +22,22 @@
   portions thereof marked with this legend must also reproduce the markings.
 """
 
+#from command_utils_base import \
+#    BasicParameter, YamlParameters, TransportCredentials
```
Remove.

phender commented on src/tests/ftest/util/dmg_utils_params.py:
```diff
@@ -22,17 +22,22 @@
+#    BasicParameter, YamlParameters, TransportCredentials
 from command_utils_base import \
-    BasicParameter, YamlParameters, TransportCredentials
+    BasicParameter, LogParameter, YamlParameters, TransportCredentials, \
+    EnvironmentVariables
```
The `EnvironmentVariables` class is not used in these changes:
```suggestion
    BasicParameter, LogParameter, TransportCredentials, YamlParameters
```

phender commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -157,6 +158,20 @@ def using_dcpm(self):
             value = self.yaml.using_dcpm
         return value
 
+    def get_interface_envs(self, index=0):
+        """Get the environment variable names and values for the interfaces.
+
+        Args:
+            index (int, optional): server index from which to obtain the
+                environment variable values. Defaults to 0.
+
+        Returns:
+            EnvironmentVariables: a dictionary of environment variable names
+                and their values extracted from the daos_server yaml
+                configuration file.
+
+        """
+        return self.yaml.get_interface_envs(index)
```
Why is this being added?  It looks like it came from an outdated copy of the file.

phender commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -312,6 +327,33 @@ def __init__(self, server_command, manager="Orterun", dmg_cfg=None):
+        # Get the values for the dmg parameters
+        self.dmg.get_params(test)
+
+    def get_interface_envs(self, index=0):
+        """Get the environment variable names and values for the interfaces.
+
+        Args:
+            index (int, optional): server index from which to obtain the
+                environment variable values. Defaults to 0.
+
+        Returns:
+            EnvironmentVariables: a dictionary of environment variable names
+                and their values extracted from the daos_server yaml
+                configuration file.
+
+        """
+        return self.manager.job.get_interface_envs(index)
```
Why is this being added?  It looks like it came from an outdated copy of the file.

phender commented on src/tests/ftest/util/agent_utils.py:
```diff
@@ -170,11 +171,15 @@ def _set_hosts(self, hosts, path, slots):
         # daos_agent processes that will be started by the command
         self.manager.job.pattern_count = len(self._hosts)
 
+
```
Remove extra inserted line.

phender commented on src/tests/ftest/launch.py:
```diff
@@ -711,6 +712,20 @@ def replace_yaml_file(yaml_file, args, tmp_dir):
 
+def generate_certs(args):
+    """Generate the certificates for the test."""
+    subprocess.call(
+        ["/usr/bin/rm", "-rf",
+         "{}/certs".format(os.environ["DAOS_TEST_LOG_DIR"])])
+    subprocess.call(["../../../../lib64/daos/certgen/gen_certificates.sh"])
+    subprocess.call(
+        ["/usr/bin/mv", "daosCA/certs",
+         "{}/".format(os.environ["DAOS_TEST_LOG_DIR"])])
+    spawn_commands(args.test_servers.split(","), "ls -la {}".format(
+        os.path.join(os.environ["DAOS_TEST_LOG_DIR"], "certs")))
+    subprocess.call(
+        ["/usr/bin/ls", "-la", os.path.join(
+            os.environ["DAOS_TEST_LOG_DIR"], "certs")])
```
I do not believe the `gen_certificates.sh` file will have permissions to write into the daos_test RPM `ftest` directory.   The ideal solution is to for `gen_certificates.sh` to support a `<destination_directory>` argument (in place of assuming `.`) and then this code could be simplified down to:
```suggestion
    certs_dir = os.path.join(os.environ["DAOS_TEST_LOG_DIR"], "certs")
    subprocess.call(["/usr/bin/rm", "-rf", certs_dir])
    subprocess.call(["../../../../lib64/daos/certgen/gen_certificates.sh", certs_dir])
```

phender commented on src/tests/ftest/util/server_utils_params.py:
```diff
@@ -32,16 +32,36 @@ class DaosServerTransportCredentials(TransportCredentials):
+        """Get certificate data.
+
+           Args:
+               name_list (list): name list.
```
```suggestion
               name_list (list): list of certificate attribute names.
```

phender commented on src/tests/ftest/util/server_utils_params.py:
```diff
@@ -32,16 +32,36 @@ class DaosServerTransportCredentials(TransportCredentials):
+
+           Returns:
+               data (dict): a dictionary of parameter directory name keys and
+               value.
```
```suggestion
                   value.
```

phender commented on src/tests/ftest/util/server_utils_params.py:
```diff
@@ -32,16 +32,36 @@ class DaosServerTransportCredentials(TransportCredentials):
+               data (dict): a dictionary of parameter directory name keys and
+               value.
+        """
+
```
```suggestion
        # Ensure the client cert directory includes the required certificate 
```

phender commented on src/tests/ftest/util/server_utils_params.py:
```diff
@@ -269,6 +289,31 @@ def update_log_files(self, control_log, helper_log, server_log):
                     "".join(log_name),
                     "server_config.server[{}].log_file".format(index))
 
+    def get_interface_envs(self, index=0):
+        """Get the environment variable names and values for the interfaces.
+
+        Args:
+            index (int, optional): server index from which to obtain the
+                environment variable values. Defaults to 0.
+
+        Returns:
+            EnvironmentVariables: a dictionary of environment variable names
+                and their values extracted from the daos_server yaml
+                configuration file.
+
+        """
+        env = EnvironmentVariables()
+        mapping = {
+            "OFI_INTERFACE": "fabric_iface",
+            "OFI_PORT": "fabric_iface_port",
+            "CRT_PHY_ADDR_STR": "provider",
+        }
+        for key, name in mapping.items():
+            value = self.server_params[index].get_value(name)
+            if value is not None:
+                env[key] = value
+
+        return env
```
Why is this being added? It looks like it came from an outdated copy of the file.

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/10/execution/node/544/log

daosbuild1 commented:
Test stage Functional_Hardware_Small completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/10/execution/node/649/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/10/execution/node/680/log

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/10/execution/node/739/log

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/10/execution/node/798/log

daosbuild1 commented on src/tests/ftest/util/apricot/apricot/test.py:
```diff
@@ -898,3 +900,4 @@ def start_additional_servers(self, additional_servers, index=0):
             self.server_managers[-1].get_config_value("filename"))
         self.server_managers[-1].verify_socket_directory(getuser())
         self.server_managers[-1].start()
+
```
(pylint-trailing-newlines) Trailing newlines

daosbuild1 commented on src/tests/ftest/launch.py:
```diff
@@ -711,6 +712,13 @@ def replace_yaml_file(yaml_file, args, tmp_dir):
     # Return the untouched or modified yaml file
     return yaml_file
 
+def generate_certs(args):
```
(pylint-unused-argument) Unused argument 'args'

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/11/execution/node/58/log

daosbuild1 commented:
Test stage Scan CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-2980/12/display/redirect

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/12/execution/node/613/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-2980/12/display/redirect

daosbuild1 commented:
Test stage Functional_Hardware_Small completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/12/execution/node/663/log

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/12/execution/node/652/log

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-2980/12/display/redirect

dinghwah commented on src/tests/ftest/launch.py:
```diff
@@ -708,10 +732,25 @@ def replace_yaml_file(yaml_file, args, tmp_dir):
     return yaml_file
 
 
+def generate_certs():
```
however, the args was not used in the function and causing pylint error.

phender replied:
My previous comment does not apply to the updated method.

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/13/execution/node/556/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/13/execution/node/620/log

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/14/execution/node/530/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/14/execution/node/620/log

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/15/execution/node/543/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/15/execution/node/721/log

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/16/execution/node/543/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/16/execution/node/620/log

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/16/execution/node/780/log

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/18/execution/node/553/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/18/execution/node/601/log

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/19/execution/node/517/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/19/execution/node/620/log

dinghwah commented:
This latest commit have addressed all of the PR comments.
(1)All tests with secure-mode by default, the failed tests with insecure-mode in test.yaml.
(2)The crt and key files generated and copied to a tmp dir DAOS_TEST_LOG_DIR for regression and PR tests.
(3)The 8 failures on the functional test are not related to this commit and tracked by:
     DAOS-5282:
     DAOS_5283:

brianjmurrell commented:
Haven't reviewed yet, but I noticed that *Test CentOS 7 RPMs* failed (silenlty :-1: ):

```
[2020-07-09T15:48:36.632Z] + skipped=0
[2020-07-09T15:48:36.632Z] + '[' '' == true ']'
[2020-07-09T15:48:36.632Z] + git show -s --format=%B
[2020-07-09T15:48:36.632Z] + grep '^Skip-test: true'
[2020-07-09T15:48:36.632Z] + '[' 0 -ne 0 ']'
[2020-07-09T15:48:36.632Z] + export DAOS_PKG_VERSION=1.1.0-26.4757.g827d527b.el7.x86_64
[2020-07-09T15:48:36.632Z] + DAOS_PKG_VERSION=1.1.0-26.4757.g827d527b.el7.x86_64
[2020-07-09T15:48:36.632Z] + ci/rpm/test_daos.sh
[2020-07-09T15:48:36.632Z] + nodelist=(${NODELIST//,/ })
[2020-07-09T15:48:36.632Z] + src/tests/ftest/config_file_gen.py -n wolf-104vm12 -a /tmp/daos_agent.yml -s /tmp/daos_server.yml
[2020-07-09T15:48:36.632Z] Traceback (most recent call last):
[2020-07-09T15:48:36.632Z]   File "src/tests/ftest/config_file_gen.py", line 27, in <module>
[2020-07-09T15:48:36.632Z]     from util.command_utils_base import CommonConfig, CommandFailure
[2020-07-09T15:48:36.632Z]   File "/data/jenkins/jenkins-1/ci_vm_wolf-104vm12/workspace/daos-stack_daos_PR-2980/src/tests/ftest/util/command_utils_base.py", line 27, in <module>
[2020-07-09T15:48:36.632Z]     from general_utils import run_command
[2020-07-09T15:48:36.632Z]   File "/data/jenkins/jenkins-1/ci_vm_wolf-104vm12/workspace/daos-stack_daos_PR-2980/src/tests/ftest/util/general_utils.py", line 32, in <module>
[2020-07-09T15:48:36.632Z]     from pathlib import Path
[2020-07-09T15:48:36.632Z] ImportError: No module named pathlib
```

It would be nice if you could fix whatever caused it to fail silently along with fixing the bug.

ravalsam commented:
pythin-pathlib is installed "Package python-pathlib-1.0.1-1.el7.noarch already installed and latest version" so import should have work from default lib path. Not sure why it happens in this PR and not others.

dinghwah commented:
@Brian, The issue for the "Test CentOS 7 RPMSs" was due to missing file of pathlib.py - DAOS-5282.
I did a search on general_uitl.py, and spoken with Samir (who did the update to this file for import pathlib),
The file should be on the package python-pathlib-1.0.1-1.el7.noarch.
There is no need for script change.

brianjmurrell commented:
> @brian, The issue for the "Test CentOS 7 RPMSs" was due to missing file of pathlib.py - DAOS-5282.

I commented on that ticket.  This regression does not appear to be happening on [latest master](https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/master/2526/pipeline) so it only stands to reason that this PR must be introducing it, or it's intermittent.  You would need to find at least one occurrence on master to assert that however.  Without any occurrences on master, we cannot assume it is an intermittent issue and have to believe that this PR is causing it, some how.

phender commented on src/tests/ftest/security/cont_create_acl.yaml:
```diff
@@ -3,15 +3,18 @@
-daos_server:
-  start:
-    insecure: True
+  transport_config:
+    allow_insecure: False
+agent_config:
+  transport_config:
+    allow_insecure: False
+dmg:
+  transport_config:
+    allow_insecure: False
```
@brianjmurrell by explicitly specifying a `allow_insecure: False` setting we ensure the test is always run, in this case, with certificates - regardless of the `launch.py` `--insecure` setting.
 
Presumably the test is written in a such a way that using certificates will cause the test to fail.  @dinghwah is that the case here?  If not I would agree with @brianjmurrell and these additions are not warranted.

brianjmurrell commented:
@dpquigl Any thoughts/comments on the process of generating a single set of certs and all of the nodes sharing/re-using them, vs. each node generating it's own unique set of certs?

daosbuild1 commented on src/tests/ftest/security/pool_connect_init.yaml:
```diff
@@ -4,6 +4,16 @@ hosts:
 timeout: 200
 server_config:
    name: daos_server
+# Uncomment following to force the use of certificates 
```
(style)  trailing whitespace

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/20/execution/node/58/log

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/21/execution/node/627/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/21/execution/node/658/log

daosbuild1 commented:
Test stage Functional_Hardware_Small completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/21/execution/node/751/log

phender commented on src/tests/ftest/launch.py:
```diff
@@ -711,6 +712,13 @@ def replace_yaml_file(yaml_file, args, tmp_dir):
+    daos_test_log_dir = os.environ["DAOS_TEST_LOG_DIR"]
+    certs_dir = os.path.join(daos_test_log_dir, "certs")
+    subprocess.call(["/usr/bin/rm", "-rf", certs_dir])
+    subprocess.call(["../../../../lib64/daos/certgen/gen_certificates.sh"])
```
This will fail in RPM installations due to the permissions of the ftest directory:
```
[pahender@wolf-10 ~]$ ll /usr/lib/daos/TESTING/
total 12
drwxr-xr-x 18 root root 4096 Jul  1 21:15 ftest
```
The ideal solution is to have `gen_certificates.sh` accept `daos_test_log_dir` as an input in which to create the files.

dpquigl commented:
> @dpquigl Any thoughts/comments on the process of generating a single set of certs and all of the nodes sharing/re-using them, vs. each node generating it's own unique set of certs?

You absolutely should not be running gen_certificates.sh more than once. I replied to this in an email weeks ago. Each run of that script generates everything. A CARoot signing key and certificate and the 3 types of key cert pairs. You should run it once and copy the necessary keys onto the nodes by type.

Compute nodes (Nodes running the agent and daos client applications) will need agent.key, agent.cert, and daosCA.crt (I think thats the name) and a properly configured agent

Administrative nodes (Any nodes running dmg) will need admin.key, admin.crt, and daosCA.crt (once again I think thats the name)

Any node running daos_server acting as a storage node will need server.crt, server.key, daosCA.crt, and under a directory called clients (or whatever its configured as in the config file) agent.crt. 

These all have to be shared. The certificate model currently is single cert per component type not cert per componet. So all admins have one cert and key, all agents have one cert and key and all servers have one cert and key.

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/22/execution/node/635/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/22/execution/node/670/log

daosbuild1 commented:
Test stage Functional_Hardware_Small completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-2980/22/testReport/(root)/

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/22/execution/node/808/log

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-2980/22/testReport/(root)/

brianjmurrell commented:
> You absolutely should not be running gen_certificates.sh more than once. I replied to this in an email weeks ago. Each run of that script generates everything. A CARoot signing key and certificate and the 3 types of key cert pairs. You should run it once and copy the necessary keys onto the nodes by type.

Interesting.  I guess I was thinking that the daos server and agent certs needed to be unique to their hostname, in the same way that web server certs are.

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
-	$CERTS/agent.crt"
+    echo "Generating Agent Certificate"
+    # Generate Private key and set its permissions
+    openssl genrsa -out $CERTS/agent.key 4096
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+    echo "Generating Agent Certificate"
+    # Generate Private key and set its permissions
+    openssl genrsa -out $CERTS/agent.key 4096
+    chmod 0400 $CERTS/agent.key
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+    openssl genrsa -out $CERTS/agent.key 4096
+    chmod 0400 $CERTS/agent.key
+    # Generate a Certificate Signing Request (CRS)
+    openssl req -new -config "$CONFIGS/agent.cnf" -key $CERTS/agent.key \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+    chmod 0400 $CERTS/agent.key
+    # Generate a Certificate Signing Request (CRS)
+    openssl req -new -config "$CONFIGS/agent.cnf" -key $CERTS/agent.key \
+        -out ${CA_HOME}/agent.csr -batch
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+    openssl req -new -config "$CONFIGS/agent.cnf" -key $CERTS/agent.key \
+        -out ${CA_HOME}/agent.csr -batch
+    # Create Certificate from request
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile $PRIVATE/daosCA.key \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+        -out ${CA_HOME}/agent.csr -batch
+    # Create Certificate from request
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile $PRIVATE/daosCA.key \
+        -cert $CERTS/daosCA.crt -policy signing_policy \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+    # Create Certificate from request
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile $PRIVATE/daosCA.key \
+        -cert $CERTS/daosCA.crt -policy signing_policy \
+        -extensions signing_agent -out $CERTS/agent.crt \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile $PRIVATE/daosCA.key \
+        -cert $CERTS/daosCA.crt -policy signing_policy \
+        -extensions signing_agent -out $CERTS/agent.crt \
+        -outdir $CERTS -in ${CA_HOME}/agent.csr -batch
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
-	rm -f test.csr
+    rm -f $CERTS/*pem
+    rm -f ${CA_HOME}/agent.csr
+    rm -f ${CA_HOME}/admin.csr
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
-	$CERTS/admin.crt"
+    echo "Generating Admin Certificate"
+    # Generate Private key and set its permissions
+    openssl genrsa -out $CERTS/admin.key 4096
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+    echo "Generating Admin Certificate"
+    # Generate Private key and set its permissions
+    openssl genrsa -out $CERTS/admin.key 4096
+    chmod 0400 $CERTS/admin.key
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+    openssl genrsa -out $CERTS/admin.key 4096
+    chmod 0400 $CERTS/admin.key
+    # Generate a Certificate Signing Request (CRS)
+    openssl req -new -config "$CONFIGS/admin.cnf" -key $CERTS/admin.key \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+    chmod 0400 $CERTS/admin.key
+    # Generate a Certificate Signing Request (CRS)
+    openssl req -new -config "$CONFIGS/admin.cnf" -key $CERTS/admin.key \
+        -out ${CA_HOME}/admin.csr -batch
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+    openssl req -new -config "$CONFIGS/admin.cnf" -key $CERTS/admin.key \
+        -out ${CA_HOME}/admin.csr -batch
+    # Create Certificate from request
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile $PRIVATE/daosCA.key \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+        -out ${CA_HOME}/admin.csr -batch
+    # Create Certificate from request
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile $PRIVATE/daosCA.key \
+        -cert $CERTS/daosCA.crt -policy signing_policy \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+    # Create Certificate from request
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile $PRIVATE/daosCA.key \
+        -cert $CERTS/daosCA.crt -policy signing_policy \
+        -extensions signing_admin -out $CERTS/admin.crt \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile $PRIVATE/daosCA.key \
+        -cert $CERTS/daosCA.crt -policy signing_policy \
+        -extensions signing_admin -out $CERTS/admin.crt \
+        -outdir $CERTS -in ${CA_HOME}/admin.csr -batch
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
-	$CERTS/server.crt"
+    echo "Generating Server Certificate"
+    # Generate Private key and set its permissions
+    openssl genrsa -out $CERTS/server.key 4096
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+    echo "Generating Server Certificate"
+    # Generate Private key and set its permissions
+    openssl genrsa -out $CERTS/server.key 4096
+    chmod 0400 $CERTS/server.key
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+    openssl genrsa -out $CERTS/server.key 4096
+    chmod 0400 $CERTS/server.key
+    # Generate a Certificate Signing Request (CRS)
+    openssl req -new -config "$CONFIGS/server.cnf" -key $CERTS/server.key \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+    chmod 0400 $CERTS/server.key
+    # Generate a Certificate Signing Request (CRS)
+    openssl req -new -config "$CONFIGS/server.cnf" -key $CERTS/server.key \
+        -out ${CA_HOME}/server.csr -batch
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+    openssl req -new -config "$CONFIGS/server.cnf" -key $CERTS/server.key \
+        -out ${CA_HOME}/server.csr -batch
+    # Create Certificate from request
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile $PRIVATE/daosCA.key \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+        -out ${CA_HOME}/server.csr -batch
+    # Create Certificate from request
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile $PRIVATE/daosCA.key \
+        -cert $CERTS/daosCA.crt -policy signing_policy \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+    # Create Certificate from request
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile $PRIVATE/daosCA.key \
+        -cert $CERTS/daosCA.crt -policy signing_policy \
+        -extensions signing_server -out $CERTS/server.crt \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile $PRIVATE/daosCA.key \
+        -cert $CERTS/daosCA.crt -policy signing_policy \
+        -extensions signing_server -out $CERTS/server.crt \
+        -outdir $CERTS -in ${CA_HOME}/server.csr -batch
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
 
 function setup_directories () {
-	mkdir -p ./daosCA/{certs,private}
+    mkdir -p $PRIVATE
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
 function setup_directories () {
-	mkdir -p ./daosCA/{certs,private}
+    mkdir -p $PRIVATE
+    mkdir -p $CERTS
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+
+function generate_ca_cnf () {
+    echo "[ ca ]" > "${CA_HOME}/ca.cnf"
+    echo "default_ca = CA_daos" >> "${CA_HOME}/ca.cnf"
```
(lint) Consider using { cmd1; cmd2; } >> file instead of individual redirects. [SC2129]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+
+    echo "Generating Test Certificate"
+    # Generate Private key and set its permissions
+    openssl genrsa -out $CERTS/test.key 4096
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
+    echo "Generating Test Certificate"
+    # Generate Private key and set its permissions
+    openssl genrsa -out $CERTS/test.key 4096
+    chmod 0400 $CERTS/test.key
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,203 @@
-	chmod 0400 $CERTS/*.key
-	chmod 0664 $CERTS/*.crt
+    chmod 0400 $CERTS/*.key
+    chmod 0664 $CERTS/*.crt
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/23/execution/node/58/log

daosbuild1 commented:
Test stage run_test.sh completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/23/execution/node/443/log

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/23/execution/node/635/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/23/execution/node/670/log

daosbuild1 commented on src/tests/ftest/launch.py:
```diff
@@ -711,6 +712,14 @@ def replace_yaml_file(yaml_file, args, tmp_dir):
+    subprocess.call(["/usr/bin/rm", "-rf", certs_dir])
+    subprocess.call(
+        ["../../../../lib64/daos/certgen/gen_certificates.sh",
+            daos_test_log_dir])
```
(pylint-bad-continuation) Wrong continued indentation (remove 3 spaces).

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
-	$CERTS/agent.crt"
+    echo "Generating Agent Certificate"
+    # Generate Private key and set its permissions
+    openssl genrsa -out ${CERTS}/agent.key 4096
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+    echo "Generating Agent Certificate"
+    # Generate Private key and set its permissions
+    openssl genrsa -out ${CERTS}/agent.key 4096
+    chmod 0400 ${CERTS}/agent.key
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+    openssl genrsa -out ${CERTS}/agent.key 4096
+    chmod 0400 ${CERTS}/agent.key
+    # Generate a Certificate Signing Request (CRS)
+    openssl req -new -config "$CONFIGS/agent.cnf" -key ${CERTS}/agent.key \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+    chmod 0400 ${CERTS}/agent.key
+    # Generate a Certificate Signing Request (CRS)
+    openssl req -new -config "$CONFIGS/agent.cnf" -key ${CERTS}/agent.key \
+        -out ${CA_HOME}/agent.csr -batch
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+    openssl req -new -config "$CONFIGS/agent.cnf" -key ${CERTS}/agent.key \
+        -out ${CA_HOME}/agent.csr -batch
+    # Create Certificate from request
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile ${PRIVATE}/daosCA.key \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+        -out ${CA_HOME}/agent.csr -batch
+    # Create Certificate from request
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile ${PRIVATE}/daosCA.key \
+        -cert ${CERTS}/daosCA.crt -policy signing_policy \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+    # Create Certificate from request
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile ${PRIVATE}/daosCA.key \
+        -cert ${CERTS}/daosCA.crt -policy signing_policy \
+        -extensions signing_agent -out ${CERTS}/agent.crt \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile ${PRIVATE}/daosCA.key \
+        -cert ${CERTS}/daosCA.crt -policy signing_policy \
+        -extensions signing_agent -out ${CERTS}/agent.crt \
+        -outdir ${CERTS} -in ${CA_HOME}/agent.csr -batch
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
-	$CERTS/admin.crt"
+    echo "Generating Admin Certificate"
+    # Generate Private key and set its permissions
+    openssl genrsa -out ${CERTS}/admin.key 4096
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+    echo "Generating Admin Certificate"
+    # Generate Private key and set its permissions
+    openssl genrsa -out ${CERTS}/admin.key 4096
+    chmod 0400 ${CERTS}/admin.key
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+    openssl genrsa -out ${CERTS}/admin.key 4096
+    chmod 0400 ${CERTS}/admin.key
+    # Generate a Certificate Signing Request (CRS)
+    openssl req -new -config "$CONFIGS/admin.cnf" -key ${CERTS}/admin.key \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+    chmod 0400 ${CERTS}/admin.key
+    # Generate a Certificate Signing Request (CRS)
+    openssl req -new -config "$CONFIGS/admin.cnf" -key ${CERTS}/admin.key \
+        -out ${CA_HOME}/admin.csr -batch
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+    openssl req -new -config "$CONFIGS/admin.cnf" -key ${CERTS}/admin.key \
+        -out ${CA_HOME}/admin.csr -batch
+    # Create Certificate from request
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile ${PRIVATE}/daosCA.key \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+        -out ${CA_HOME}/admin.csr -batch
+    # Create Certificate from request
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile ${PRIVATE}/daosCA.key \
+        -cert ${CERTS}/daosCA.crt -policy signing_policy \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+    # Create Certificate from request
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile ${PRIVATE}/daosCA.key \
+        -cert ${CERTS}/daosCA.crt -policy signing_policy \
+        -extensions signing_admin -out ${CERTS}/admin.crt \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile ${PRIVATE}/daosCA.key \
+        -cert ${CERTS}/daosCA.crt -policy signing_policy \
+        -extensions signing_admin -out ${CERTS}/admin.crt \
+        -outdir ${CERTS} -in ${CA_HOME}/admin.csr -batch
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
-	$CERTS/server.crt"
+    echo "Generating Server Certificate"
+    # Generate Private key and set its permissions
+    openssl genrsa -out ${CERTS}/server.key 4096
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+    echo "Generating Server Certificate"
+    # Generate Private key and set its permissions
+    openssl genrsa -out ${CERTS}/server.key 4096
+    chmod 0400 ${CERTS}/server.key
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+    openssl genrsa -out ${CERTS}/server.key 4096
+    chmod 0400 ${CERTS}/server.key
+    # Generate a Certificate Signing Request (CRS)
+    openssl req -new -config "$CONFIGS/server.cnf" -key ${CERTS}/server.key \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+    chmod 0400 ${CERTS}/server.key
+    # Generate a Certificate Signing Request (CRS)
+    openssl req -new -config "$CONFIGS/server.cnf" -key ${CERTS}/server.key \
+        -out ${CA_HOME}/server.csr -batch
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
 
 function setup_directories () {
-	mkdir -p ./daosCA/{certs,private}
+    mkdir -p ${PRIVATE}
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
 function setup_directories () {
-	mkdir -p ./daosCA/{certs,private}
+    mkdir -p ${PRIVATE}
+    mkdir -p ${CERTS}
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+        -out ${CA_HOME}/server.csr -batch
+    # Create Certificate from request
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile ${PRIVATE}/daosCA.key \
+        -cert ${CERTS}/daosCA.crt -policy signing_policy \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+    # Create Certificate from request
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile ${PRIVATE}/daosCA.key \
+        -cert ${CERTS}/daosCA.crt -policy signing_policy \
+        -extensions signing_server -out ${CERTS}/server.crt \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile ${PRIVATE}/daosCA.key \
+        -cert ${CERTS}/daosCA.crt -policy signing_policy \
+        -extensions signing_server -out ${CERTS}/server.crt \
+        -outdir ${CERTS} -in ${CA_HOME}/server.csr -batch
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+
+function generate_ca_cnf () {
+    echo "[ ca ]" > "${CA_HOME}/ca.cnf"
+    echo "default_ca = CA_daos" >> "${CA_HOME}/ca.cnf"
```
(lint) Consider using { cmd1; cmd2; } >> file instead of individual redirects. [SC2129]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+
+    echo "Generating Test Certificate"
+    # Generate Private key and set its permissions
+    openssl genrsa -out ${CERTS}/test.key 4096
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+    echo "Generating Test Certificate"
+    # Generate Private key and set its permissions
+    openssl genrsa -out ${CERTS}/test.key 4096
+    chmod 0400 ${CERTS}/test.key
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile "${PRIVATE}/daosCA.key" \
+        -cert "${CERTS}/daosCA.crt" -policy signing_policy \
+        -extensions signing_test -out "${CERTS}/test.crt" \
+        -outdir ${CERTS} -in "${CA_HOME}/test.csr" -batch
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+    openssl req -new -config "$CONFIGS/server.cnf" -key ${CERTS}/server.key \
+        -out ${CA_HOME}/server.csr -batch
+    # Create Certificate from request
+    openssl ca -config "$CA_HOME/ca.cnf" -keyfile ${PRIVATE}/daosCA.key \
```
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/24/execution/node/59/log

daosbuild1 commented on utils/certs/gen_certificates.sh:
```diff
@@ -38,146 +38,210 @@
+
+function generate_ca_cnf () {
+    echo "[ ca ]" > "${CA_HOME}/ca.cnf"
+    echo "default_ca = CA_daos" >> "${CA_HOME}/ca.cnf"
```
(lint) Consider using { cmd1; cmd2; } >> file instead of individual redirects. [SC2129]

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/26/execution/node/59/log

daosbuild1 commented:
Test stage Build RPM on Leap 15 completed with status UNSTABLE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/26/execution/node/242/log

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/26/execution/node/629/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/26/execution/node/673/log

daosbuild1 commented:
Test stage Functional_Hardware_Small completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-2980/26/testReport/(root)/

daosbuild1 commented on src/tests/ftest/unittest/unittest.py:
```diff
@@ -22,8 +22,34 @@
+
+    file_path = list(Path(basepath).glob('**/{0}'.format(bin_name)))
+    if not file_path:
+        raise OSError(ENOENT, "File {0} not found inside {1} Directory"
```
(pylint-undefined-variable) Undefined variable 'ENOENT'

daosbuild1 commented on src/tests/ftest/unittest/unittest.py:
```diff
@@ -22,8 +22,34 @@
+        OSError: If failed to find the bin_name file
+    """
+    with open('../../.build_vars.json') as json_file:
+        build_paths = json.load(json_file)
```
(pylint-undefined-variable) Undefined variable 'json'

daosbuild1 commented on src/tests/ftest/unittest/unittest.py:
```diff
@@ -22,8 +22,34 @@
+    """
+    with open('../../.build_vars.json') as json_file:
+        build_paths = json.load(json_file)
+    basepath = os.path.normpath(build_paths['PREFIX'] + "/../{0}"
```
(pylint-undefined-variable) Undefined variable 'os'

daosbuild1 commented on src/tests/ftest/control/dmg_storage_query.py:
```diff
@@ -27,6 +27,7 @@
 import re
 from command_utils import CommandFailure
 from control_test_base import ControlTestBase
+from apricot import TestWithServers, skipForTicket
```
(pylint-unused-import) Unused TestWithServers imported from apricot

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/27/execution/node/58/log

brianjmurrell commented on src/tests/ftest/control/dmg_storage_query.py:
```diff
@@ -60,6 +61,7 @@ def check_dev_state(self, devs_info, state):
         if err:
             self.fail("Found device(s) in bad state: {}".format(err))
 
+    @skipForTicket("DAOS-4286")
```
DAOS-4286 is **Resolved**.  You can't skip a test for a ticket that has been resolved.

dinghwah replied:
@Brian Thanks for the review and comment, the failures should be DAOS-5320.

brianjmurrell commented on src/tests/ftest/control/dmg_storage_query.py:
```diff
@@ -82,6 +84,7 @@ def test_dmg_storage_query_devices(self):
         if self.targets != targets:
             self.fail("Wrong number of targets found: {}".format(targets))
 
+    @skipForTicket("DAOS-4286")
```
DAOS-4286 is **Resolved**.  You can't skip a test for a ticket that has been resolved.

brianjmurrell commented on src/tests/ftest/control/dmg_storage_query.py:
```diff
@@ -118,6 +121,7 @@ def test_dmg_storage_query_pools(self):
         no_pool_info = self.get_pool_info()
         self.assertFalse(no_pool_info, "No pools should be detected.")
 
+    @skipForTicket("DAOS-4286")
```
DAOS-4286 is **Resolved**.  You can't skip a test for a ticket that has been resolved.

brianjmurrell commented on src/tests/ftest/control/dmg_storage_query.py:
```diff
@@ -169,6 +173,7 @@ def test_dmg_storage_query_device_health(self):
         if err:
             self.fail("Health info not as expected: {}".format(err))
 
+    @skipForTicket("DAOS-4286")
```
DAOS-4286 is **Resolved**.  You can't skip a test for a ticket that has been resolved.

brianjmurrell commented on src/tests/ftest/security/pool_security_groups.yaml:
```diff
@@ -5,19 +5,18 @@ hosts:
-    - key: ~/.daos/daos_agent.key
-    - server_name: server
-timeout: 600
+timeout: 1200
```
Please don't resolve conversations that were started by somebody else.  It is up to the person that starts a conversation as to whether it's resolved or not.

That said, you resolved this without answering the question.  Why does it need doubling?

dinghwah commented:
From Brian's comment: 
Are the certs being copied to the system/production location? It appears they are being copied to the /var/tmp/daos_testing log directory. Why not use the system/production certificate directory so that there is less effort required when it comes time to starting servers with systemd, etc.?
--> We have updated gen_certificates.sh more flexible to allow users or new applications or tests to store all the needed cert-files to any location.
      For CI and PR functional tests, we are storing the cert-files to a tmp directory (/var/tmp/daos_testing/daosCA), used for different servers, agents and admin-users for the needed files. Credential files also needed to be cleaned after functional tests completed.

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/28/execution/node/613/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/28/execution/node/692/log

dinghwah commented:
@Brian, forgot to mention, src/tests/ftest/security/pool_security_groups.yaml  -timeout 600 has already been restored.

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-2980/28/testReport/(root)/

brianjmurrell commented on Jenkinsfile:
```diff
@@ -1201,6 +1202,7 @@ pipeline {
                     }
                     steps {
                         testRpm inst_repos: el7_daos_repos(),
+                                inst_rpms: test_rpm_rpms,
```
This is not an avocado driven test.

phender replied:
It is not, but the `config_gile_gen.py` script it uses to generate the configuration files for the `daos_agent` and `daos_server` commands it uses to test the DAOS RPMs (as part of this PR) imports (but does not really need to use for the purpose of testing the DAOS RPMs) the avocado.utils python library.

We can avoid edits to this file and the import problems (related to the current `command_utils_base.py` updates), by instead moving the code in `command_utils_base.py` that imports `general_utils.py` to the `command_utils.py` file (not needed/imported by `config_fiel_gen.py`).

brianjmurrell replied:
> It is not, but the `config_gile_gen.py` script it uses to generate the configuration files for the `daos_agent` and `daos_server` commands it uses to test the DAOS RPMs (as part of this PR) imports (but does not really need to use for the purpose of testing the DAOS RPMs) the avocado.utils python library.

I see.  It sounds like an avocado-independent version of this tool needs to be written perhaps then, yes?
 
> We can avoid edits to this file and the import problems (related to the current `command_utils_base.py` updates), by instead moving the code in `command_utils_base.py` that imports `general_utils.py` to the `command_utils.py` file (not needed/imported by `config_fiel_gen.py`).

I guess this is accomplishing the above suggestion then?

phender replied:
> I guess this is accomplishing the above suggestion then?
yes
 
> In any case, the avocado RPMs being added to `inst_rpms` here are already installed on the test node, so there is no need to specify them here also.
Although this is now inconsequential with the removal of the `import general_utils.py` in `command_utils_base.py`, doesn't the python `ImportError` for `pathlib` indicate otherwise?  Presumably there is nothing special needed in terms of PYTHONPATH setup for `pathlib` - we just install the RPM and it works.

brianjmurrell replied:
I don't know why `pathlib` is not being found, but it is installed even before this change as we can see in the most recent run:

```
[2020-07-18T21:54:35.394Z] + yum -y install CUnit python36-PyYAML python36-nose python36-pip valgrind python36-paramiko python2-avocado python2-avocado-plugins-output-html python2-avocado-plugins-varianter-yaml-to-mux libcmocka python-pathlib python2-numpy git python2-clustershell golang-bin ipmctl ndctl
[2020-07-18T21:54:35.394Z] Loaded plugins: fastestmirror
[2020-07-18T21:54:35.394Z] Loading mirror speeds from cached hostfile
[2020-07-18T21:54:35.394Z]  * base: mirrors.sonic.net
[2020-07-18T21:54:35.394Z]  * epel: mirror.chpc.utah.edu
[2020-07-18T21:54:35.394Z]  * extras: mirrors.gigenet.com
[2020-07-18T21:54:35.394Z]  * updates: mirror.chpc.utah.edu
[2020-07-18T21:54:35.394Z] Package CUnit-2.1.3-8.el7.x86_64 already installed and latest version
[2020-07-18T21:54:35.394Z] Package python36-PyYAML-3.13-1.el7.x86_64 already installed and latest version
[2020-07-18T21:54:35.394Z] Package python36-nose-1.3.7-4.el7.noarch already installed and latest version
[2020-07-18T21:54:35.394Z] Package python3-pip-9.0.3-7.el7_7.noarch already installed and latest version
[2020-07-18T21:54:35.394Z] Package 1:valgrind-3.15.0-11.el7.x86_64 already installed and latest version
[2020-07-18T21:54:35.394Z] Package python36-paramiko-2.1.1-0.10.el7.noarch already installed and latest version
[2020-07-18T21:54:35.394Z] Package python2-avocado-52.1-6.el7.noarch already installed and latest version
[2020-07-18T21:54:35.394Z] Package python2-avocado-plugins-output-html-52.1-6.el7.noarch already installed and latest version
[2020-07-18T21:54:35.394Z] Package python2-avocado-plugins-varianter-yaml-to-mux-52.1-6.el7.noarch already installed and latest version
[2020-07-18T21:54:35.394Z] Package libcmocka-1.1.5-1.el7.x86_64 already installed and latest version
[2020-07-18T21:54:35.394Z] Package python-pathlib-1.0.1-1.el7.noarch already installed and latest version
[2020-07-18T21:54:35.394Z] Package 1:python2-numpy-1.7.1-0.el7.noarch already installed and latest version
[2020-07-18T21:54:35.394Z] Package git-1.8.3.1-23.el7_8.x86_64 already installed and latest version
[2020-07-18T21:54:35.394Z] Package python2-clustershell-1.8.3-1.el7.noarch already installed and latest version
[2020-07-18T21:54:35.394Z] Package golang-bin-1.13.11-1.el7.x86_64 already installed and latest version
[2020-07-18T21:54:35.394Z] Package ipmctl-02.00.00.3764-1.el7.x86_64 already installed and latest version
[2020-07-18T21:54:35.394Z] Package ndctl-65-5.el7.x86_64 already installed and latest version
[2020-07-18T21:54:35.394Z] Nothing to do
```

phender commented on src/tests/ftest/util/server_utils.py:
```diff
@@ -327,6 +341,12 @@ def prepare(self, storage=True):
         self.manager.job.create_yaml_file()
 
+        # Copy certificates
+        self.manager.job.yaml.copy_certificates(
```
```suggestion
        self.manager.job.copy_certificates(
```

phender commented on src/tests/ftest/util/agent_utils.py:
```diff
@@ -175,6 +176,9 @@ def start(self):
             "<AGENT> Starting daos_agent on %s with %s",
             self._hosts, self.manager.command)
+        # Copy certificates
+        self.manager.job.yaml.copy_certificates(
```
```suggestion
        self.manager.job.copy_certificates(
```

phender commented on src/tests/ftest/util/server_utils_params.py:
```diff
@@ -24,24 +24,49 @@
+        self.cert = LogParameter(log_dir, None, "server.crt")
+        self.key = LogParameter(log_dir, None, "server.key")
+
+    def _get_certificate_data(self, name_list):
```
```suggestion
    def get_certificate_data(self, name_list):
```

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/29/execution/node/612/log

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/31/execution/node/587/log

phender commented on src/tests/ftest/util/command_utils.py:
```diff
@@ -685,6 +686,35 @@ def _get_result(self):
+                    for file_name in data[name]:
+                        src_file = os.path.join(source, file_name)
+                        dst_file = os.path.join(name, file_name)
+                        run_command(
+                            "clush -S -v -w {} --copy {} --dest {}".format(
+                                ",".join(hosts), src_file, dst_file))
```
To support testing with invalid certificate filenames, this code should log a warning instead of raising an exception:
```suggestion
                        result = run_command(
                            "clush -S -v -w {} --copy {} --dest {}".format(
                                ",".join(hosts), src_file, dst_file),
                            raise_exception=False)
                        if result.exit_status != 0:
                            self.log.info("WARNING: failure copying '%s' to '%s' on %s", src_file, dst_file, hosts)
```

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/31/execution/node/851/log

daosbuild1 commented on src/tests/ftest/util/command_utils.py:
```diff
@@ -685,6 +686,40 @@ def _get_result(self):
+                        if result.exit_status != 0:
+                            self.log.info(
+                                "WARNING: failure copying '%s' to '%s' on %s",
+                            src_file, dst_file, hosts)
```
(pylint-bad-continuation) Wrong hanging indentation (add 4 spaces).

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/32/execution/node/58/log

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/33/execution/node/708/log

daosbuild1 commented:
Test stage Functional_Hardware_Small completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-2980/33/testReport/(root)/

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/33/execution/node/795/log

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-2980/33/testReport/(root)/

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/33/execution/node/783/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/35/execution/node/690/log

daosbuild1 commented:
Test stage Functional_Hardware_Small completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-2980/35/testReport/(root)/

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/35/execution/node/845/log

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/35/execution/node/788/log

dinghwah commented:
(1)Thanks to Phil and Brian for the review and comments, all the comments have been addressed.
(2)I have pushed the PR with CI weekly test tags, and triaged the test-results of 28 failures which are irrelevant to the certificates enabled.
(3)Test-results are consistent with the latest CI weekly test-results #94.  
     https://build.hpdd.intel.com/job/daos-stack/job/daos/job/weekly-testing/94/#showFailuresLink.
(4)The issue of failure in "Test CentOS 7 RPMs" has been resolved.
(5)Tests with certificates enable are stable.

dinghwah commented on src/tests/ftest/control/daos_agent_config.yaml:
```diff
@@ -86,46 +93,47 @@ agent_config_val: !mux
+#    config_val:
+#       - "cert"
+#       - /tmp/does_not_exist.ext
+#       - "PASS"
```
I have updated the script and uncommented the skipped 6 testcases, they are working now.

phender commented:
Hi Ding, I still have one outstanding comment against the changes to `src/tests/ftest/control/daos_control_config.yaml` (just force insecure mode and restore the original PASS/FAIL criteria) and you still need a normal CI run (default pragmas) against the latest commit as it made changes to `src/tests/ftest/util/command_utils.py`.

daosbuild1 commented:
Test stage Functional_Hardware_Small completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-2980/37/testReport/(root)/

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/37/execution/node/843/log

dinghwah commented:
The 1 failure in the functional test was due to a known issue DAOS-5311.

mjmac commented:
I am reluctant to land this without seeing a successful run of the Functional Hardware Large suite, given that this PR involves making major changes to the tests. Efforts are being made to resolve the logging problem that resulted in this failure, so I recommend waiting until master is expected to be stable again, merging in the fixes, and then re-running the tests in this PR to verify that there are no regressions.

In the meantime, please update the subject and first comment to conform to DAOS commit message standards: https://wiki.hpdd.intel.com/display/DC/Commit+Comments

mjmac commented on src/tests/ftest/control/daos_admin_privileged.yaml:
```diff
@@ -13,3 +13,11 @@ server_config:
     # bdev_list: ["aaaa:aa:aa.a","bbbb:bb:bb.b"]
     scm_class: dcpm
     scm_list: ["/dev/pmem0"]
+  transport_config:
+    allow_insecure: True
+agent_config:
+  transport_config:
+    allow_insecure: True
+dmg:
+  transport_config:
+    allow_insecure: True
```
Why are these tests allowed to run in insecure mode? The commit message does not include any information about why some tests have it enabled and some do not.

saurabhtandan commented on src/tests/ftest/security/pool_connect_init.yaml:
```diff
@@ -4,6 +4,16 @@ hosts:
+#     allow_insecure: False
+#dmg:
+#   transport_config:
+#     allow_insecure: False
```
So what's happening in this test? This has neither allow_insecure as True or false. What is the default? Other yaml files above, either have it mentioned as True of false. SO just trying to understand what is happening when it is not mentioned.

dinghwah replied:
Certificates will be enabled by default including this test, the commented-out lines providing a reference in case there is a need to run with insecure-mode.

daosbuild1 commented:
Test stage Functional_Hardware_Small completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-2980/38/testReport/(root)/

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/38/execution/node/842/log

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/38/execution/node/803/log

daosbuild1 commented:
Test stage Functional_Hardware_Small completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-2980/39/testReport/(root)/

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-2980/39/execution/node/750/log

dinghwah commented:
2 failures were due to known failures DAOS-5215 and DAOS-4630.
